src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+{
+    instruction_ref ins0               = instrs.at(0);
+    instruction_ref base               = root[ins0];
+    std::vector<std::size_t> base_lens = base->get_shape().lens();
+    int axis                           = find_axis(ins0, base, root_axis);
+
+    int sum      = 0;
+    int base_sum = 0;
+    for(auto&& ins : instrs)
+    {
+        sum += ins->get_shape().lens().at(axis);
+        base_sum += root[ins]->get_shape().lens().at(root_axis);
+    }
+
+    base_lens[root_axis] = base_sum;
+
+    if(ins0->outputs().size() != 1)
+        MIGRAPHX_THROW("unexpected output size");
+    instruction_ref output            = ins0->outputs().at(0);
+    shape s                           = ins0->get_shape();
+    std::vector<std::size_t> new_lens = s.lens();
+    new_lens[axis]                    = sum;
+
+    if(ins0->name() == "@literal")
+    {
+        // concat literals.
+        unsigned long long unit_slice   = 1;
+        int ndx                         = 0;
+        unsigned long long new_elements = 1;
+
+        for(auto&& len : s.lens())
+        {
+            if(ndx > axis)
+                unit_slice *= len;
+            if(ndx != axis)
+                new_elements *= len;
+            ndx++;
+        }
+        new_elements *= sum;
+        unsigned type_size             = s.type_size();
+        unsigned long long total_bytes = new_elements * type_size;
+        std::shared_ptr<char> input    = make_shared_array<char>(total_bytes);
+        std::vector<unsigned long long> bytes_per_slice;
+        unsigned long long unit_slice_bytes = unit_slice * type_size;
+
+        std::transform(instrs.begin(),
+                       instrs.end(),
+                       std::back_inserter(bytes_per_slice),
+                       [&](auto&& d) -> unsigned long long {
+                           return d->get_shape().lens().at(axis) * unit_slice_bytes;
+                       });
+
+        unsigned copy_bytes = 0;
+        int slice_ndx       = 0;
+        char* dst           = input.get();
+        while(copy_bytes < total_bytes)
+        {
+            unsigned ins_ndx = 0;
+            for(auto&& ins : instrs)
+            {
+                unsigned long long bytes = bytes_per_slice[ins_ndx];
+                const char* src          = &(ins->get_literal().data()[slice_ndx * bytes]);
+                std::copy(src, src + bytes, dst);
+                dst += bytes;
+                copy_bytes += bytes;
+                ins_ndx++;
+            }
+            slice_ndx++;
+        }
+        shape new_shape{s.type(), new_lens};
+        auto new_literal = p_program->add_literal(literal(new_shape, input.get()));
+        instruction::replace_argument(output, ins0, new_literal, false);
+    }
+    else
+    {
+        ins0->set_shape({ins0->get_shape().type(), new_lens});
+        if(ins0->name() == "broadcast")
+        {
+            // workaround for a bad practice: broadcast has a broadcast_shape field.
```

Comment:
This is no longer the case and can be removed.

Reply:
still has a broadcast_lens field.  This kind of caching a field that is derived from recomputation is strange.

Reply:
Removing this triggers a program verification error.  In the past, it is shape that is cached.  Shucai changed it to remove the "type" part.  It is sufficient for his retyping work.   But insufficient if dimensions are changed, which is the case here.